### PR TITLE
API-1835: add mutation tracking to the fake client

### PR DESCRIPTION
**Pull request by deads2k**
Adds the ability to track mutations to our manifest client with unit tests.  We can build on this for comparing predicted versus actual mutations.

### CODE REVIEW

File: pkg/manifestclient/write_roundtripper.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	bodyObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, bodyContent)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode body: %w", err)
+	}
+	bodyYAMLBytes, err := yaml.Marshal(bodyObj.(*unstructured.Unstructured).Object)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encode body: %w", err)
+	}
+	bodyJSONBytes, err := json.Marshal(bodyObj.(*unstructured.Unstructured).Object)
```
Comment (p0lyn0mial) [inline pkg/manifestclient/write_roundtripper.go:106]: why ? cannot you just return `bodyObj` ?

File: pkg/manifestclient/write_roundtripper.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		bodyContent, err = io.ReadAll(req.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read body: %w", err)
+		}
+	}
+	bodyObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, bodyContent)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode body: %w", err)
+	}
+	bodyYAMLBytes, err := yaml.Marshal(bodyObj.(*unstructured.Unstructured).Object)
```
Comment (p0lyn0mial) [inline pkg/manifestclient/write_roundtripper.go:102]: i assume that yaml encoding is just for convenience.

File: pkg/manifestclient/readwrite_roundtripper.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// Enter here and call `NewForConfigAndClient(&rest.Config{}, httpClient)`
```
Comment (p0lyn0mial) [inline pkg/manifestclient/readwrite_roundtripper.go:11]: ?

File: pkg/manifestclient/readwrite_roundtripper.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+type MutationTrackingClient interface {
+	GetHTTPClient() *http.Client
+	GetMutations() *AllActionsTracker
+}
+
+func (rt *readWriteRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	switch req.Method {
+	case "GET", "HEAD":
+		return rt.readDelegate.RoundTrip(req)
```
Comment (p0lyn0mial) [inline pkg/manifestclient/readwrite_roundtripper.go:80]: have you pushed all your changes? i cannot find the readDelegate impl.

File: pkg/manifestclient/mutation_tracker.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	Name               string
+	SerializedRequests []SerializedRequest
+}
+
+type SerializedRequest struct {
+	Options []byte
+	Body    []byte
+}
+
+var count = 1
```
Comment (p0lyn0mial) [inline pkg/manifestclient/mutation_tracker.go:59]: unused?

File: pkg/manifestclient/write_roundtripper.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	requestInfo, err := mrt.requestInfoResolver.NewRequestInfo(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading requestInfo: %w", err)
+	}
+
+	if !requestInfo.IsResourceRequest {
+		return nil, fmt.Errorf("non-resource requests are not supported by this implementation")
+	}
+	if len(requestInfo.Subresource) != 0 && requestInfo.Subresource != "status" {
+		return nil, fmt.Errorf("subresource %v is not supported by this implementation", requestInfo.Subresource)
```
Comment (p0lyn0mial) [inline pkg/manifestclient/write_roundtripper.go:70]: why ? what about `scale` ?

File: pkg/manifestclient/write_roundtripper.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+		Name:      requestInfo.Name,
+	}
+	if action == ActionCreate {
+		// in this case, the name isn't in the URL, it's in the body
+		metadata.Name = bodyObj.(*unstructured.Unstructured).GetName()
+	}
+
+	mrt.actionTracker.AddRequest(metadata, serializedRequest)
+
+	return bodyJSONBytes, nil
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/manifestclient/write_roundtripper.go:132]: > does it mean that the users don't make use the returned objects ?

Will document.  Correct, they cannot use the output.

[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/manifestclient/write_roundtripper.go:132]: > Will document. Correct, they cannot use the output.

Could I make it more obvious with some different kind of return?

[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/manifestclient/write_roundtripper.go:132]: I'm making it very obvious with the return value.

File: pkg/manifestclient/write_roundtripper.go (hunk 1 above)
Comment (deads2k) [reply pkg/manifestclient/write_roundtripper.go:106]: > why ? cannot you just return `bodyObj` ?

roundtrippers return bytes.  if we desire a refactor at some point, maybe.  But I think this gives us flexibility where we want it.

File: pkg/manifestclient/readwrite_roundtripper.go (hunk 4 above)
Comment (deads2k) [reply pkg/manifestclient/readwrite_roundtripper.go:80]: > have you pushed all your changes? i cannot find the readDelegate impl.

It merged a couple months back.

File: pkg/manifestclient/write_roundtripper.go (hunk 6 above)
Comment (deads2k) [reply pkg/manifestclient/write_roundtripper.go:70]: > why ? what about `scale` ?

If we need it, we'll add it.  For now getting this off the ground with enough to start using.

**Comment by p0lyn0mial** [comment]
lgtm

let's give it a try.

**Pull request merged**